src/dbnode/storage/series/reader.go

```diff
@@ -119,88 +119,174 @@ func (r *Reader) readersWithBlocksMapAndBuffer(
 		seriesBlocks, seriesBuffer, nsCtx)
 }
 
+// BlockReaderIter provides an Iterator interface to a collection of BlockReaders.
+//
+// The Iterator allows disk read to be lazily requested when Next() is called. This allows the system to limit an
+// expensive Query before it issues all its disk reads.
+//
+// The Iterator does not hold any locks so it's safe to pause or even abandon iterating through the results. Any
+// in-memory blocks (cache, series buffer) are eagerly loaded to avoid holding read locks while iterating.
+//
+// This iterator is not thread safe and should be only be used by a single go routine.
+type BlockReaderIter interface {
+	// Next requests the next BlockReaders, returning if any more existed.
+	// The BlockReaders are available with Current.
+	// If an error occurs, false is returned and it is available with Err.
+	Next(ctx context.Context) bool
+
+	// Current returns the current set of BlockReaders for a given blockStart time, guaranteed to be not empty.
+	// Results are returned in block start time asc ordering. Within the same block start time there is no guaranteed
+	// ordering.
+	Current() []xio.BlockReader
+
+	// Err is non-nil if an error occurred when calling Next.
+	Err() error
+
+	// ToSlices eagerly loads all BlockReaders into the legacy slices of slices for backward compatibility.
+	// TODO: remove this and convert everything to the iterator pattern.
+	ToSlices(ctx context.Context) ([][]xio.BlockReader, error)
+}
+
+type blockReaderIter struct {
+	start     time.Time
+	end       time.Time
+	blockSize time.Duration
+	reader    *Reader
+	nsCtx     namespace.Context
+	cached    []xio.BlockReader
+	buffer    [][]xio.BlockReader
+
+	blockAt time.Time
+	curr    []xio.BlockReader
+	err     error
+}
+
+func (i *blockReaderIter) Err() error {
+	return i.err
+}
+
+func (i *blockReaderIter) Current() []xio.BlockReader {
+	return i.curr
+}
+
 // nolint: gocyclo
+func (i *blockReaderIter) Next(ctx context.Context) bool {
+	if i.blockAt.IsZero() {
+		i.blockAt = i.start
+	}
+	i.curr = make([]xio.BlockReader, 0)
+	for !i.blockAt.After(i.end) && len(i.curr) == 0 {
```

Comment:
probably worth a close eye on this change, since the logic was kind of tricky to preserve the existing ordering by block start time. had to interleave in memory blocks with disk requests.